pkg/controllers/user/nodesyncer/cordonfieldssyncer.go

```diff
@@ -47,31 +55,79 @@ func (m *NodesSyncer) syncCordonFields(key string, obj *v3.Node) error {
 }
 
 func (d *NodeDrain) drainNode(key string, obj *v3.Node) error {
-	if obj == nil || obj.DeletionTimestamp != nil || obj.Spec.DesiredNodeUnschedulable != "drain" {
+	if obj == nil || obj.DeletionTimestamp != nil || obj.Spec.DesiredNodeUnschedulable == "" {
 		return nil
 	}
-	kubeConfig, err := d.getKubeConfig()
-	if err != nil {
-		return err
+
+	if obj.Spec.DesiredNodeUnschedulable != "drain" && obj.Spec.DesiredNodeUnschedulable != "cancel" {
+		return nil
+	}
+
+	if obj.Spec.DesiredNodeUnschedulable == "drain" {
+		nodeMapLock.Lock()
+		if _, ok := d.nodesToContext[obj.Name]; ok {
+			nodeMapLock.Unlock()
+			return nil
+		}
+		ctx, cancel := context.WithCancel(context.Background())
+		d.nodesToContext[obj.Name] = cancel
+		nodeMapLock.Unlock()
+		go d.drain(ctx, obj)
+
+	} else if obj.Spec.DesiredNodeUnschedulable == "cancel" {
+		nodeMapLock.Lock()
```

Comment:
have you considered using sync.map? https://medium.com/@deckarep/the-new-kid-in-town-gos-sync-map-de24a6bf7c2c